src/persist/src/file.rs

```diff
@@ -327,8 +328,92 @@ impl Blob for FileBlob {
         Ok(Some(buf))
     }
 
+    async fn list_keys(&self) -> Result<Vec<String>, Error> {
+        let base_dir = match &self.base_dir {
+            Some(base_dir) => base_dir.canonicalize()?,
+            None => return Err(Error::from("FileBlob unexpectedly closed")),
+        };
+        let mut ret = vec![];
+
+        for entry in fs::read_dir(&base_dir)? {
+            let entry = entry?;
+            let path = entry.path().canonicalize()?;
+
+            if !path.is_file() {
+                // Ignore '.' and '..' directory entries if they come up.
+                if path == base_dir {
+                    continue;
+                } else if let Some(parent) = base_dir.parent() {
+                    if path == parent {
+                        continue;
+                    }
+                } else {
+                    return Err(Error::from(format!(
+                        "unexpectedly found directory while iterating through FileBlob: {}",
+                        path.display()
+                    )));
+                }
+            }
+
+            // The file name is guaranteed to be non-None iff the path is a
+            // normal file.
+            let file_name = path.file_name();
+            if let Some(name) = file_name {
+                let name = name.to_str();
+                if let Some(name) = name {
+                    ret.push(name.to_owned());
+                }
+            }
+        }
+        Ok(ret)
+    }
+}
+
+#[async_trait]
+impl BlobRead for FileBlob {
+    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>, Error> {
+        self.read.get(key).await
+    }
+
+    async fn list_keys(&self) -> Result<Vec<String>, Error> {
+        self.read.list_keys().await
+    }
+}
+
+#[async_trait]
+impl Blob for FileBlob {
+    type Config = FileBlobConfig;
```

Comment:
Hmm I think these associated types make it difficult to use downcast_ref to call close on a BlobRead that might be a Blob

Reply:
yeah decided to move close to BlobRead and impl it differently between e.g. FileBlob and FileBlobRead, even if that is a little wonky. added a comment on the rustdoc for close